### ISSUE TITLE
math.Cos(1.23e300) != 0.251533

### ISSUE THREAD

**Issue by elliotchance** [golang/go#20539]
### What version of Go are you using (`go version`)?

go version go1.7.5 darwin/amd64

### What operating system and processor architecture are you using (`go env`)?

GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/elliot/go/"
GORACE=""
GOROOT="/usr/local/go"
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
CC="clang"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/c7/q_xt9q_j0_l99wstm08c1_280000gn/T/go-build540953956=/tmp/go-build -gno-record-gcc-switches -fno-common"
CXX="clang++"
CGO_ENABLED="1"

### What did you do?

https://play.golang.org/p/DaHoTvn29Z

### What did you expect to see?

In C (clang) on the same machine:

```c
cos(1.23e300) == 0.251533
```

I realise such a large value would be an extreme and almost certainly incorrect value for `cos()`, but theoretically any input finite value should return a finite value.

### What did you see instead?

-inf

**Comment by elliotchance** [comment]
A similar problem happens with `math.Cosh`:

```go
	// Outputs 0, should be: -Inf
	fmt.Println(math.Cosh(1.23e300))
```

https://play.golang.org/p/ah5FAF5tVI

**Comment by elliotchance** [comment]
Also, `math.Exp`. E to the power of an very big number should probably return +Inf, instead we get 0:

```go
	fmt.Println(math.Exp(1.23e300))
```

**Comment by ALTree** [comment]
The `Exp` problem was fixed on tip by f44e58703115af61e7b03416273031d788c076f1
The `Sin` / `Cos` problem is #6794

**Comment by ALTree** [comment]
Closing this as dup of an older issue (#6794)

**Comment by Konstantin8105** [comment]
In my point of view, if you try to calculate the Cos(1e300), then correct answer is panic or error, because the angle is 100000000000... and result is just from 1 to -1.
But we know only 16 number with float64(see https://en.wikipedia.org/wiki/Double-precision_floating-point_format) after point.

**Comment by elliotchance** [comment]
@Konstantin8105 how did you come up with the result Cos(1e300) = 100000000000?

**Comment by Konstantin8105** [comment]
At the first I want to show example from fumction Sqrt, see https://golang.org/src/math/sqrt.go : 
```golang
// Sqrt returns the square root of x.
  //
  // Special cases are:
  //	Sqrt(+Inf) = +Inf
  //	Sqrt(±0) = ±0
  //	Sqrt(x < 0) = NaN
  //	Sqrt(NaN) = NaN
  func Sqrt(x float64) float64
...
func sqrt(x float64) float64 {
  	// special cases
  	switch {
  	case x == 0 || IsNaN(x) || IsInf(x, 1):
  		return x
  	case x < 0:
  		return NaN()
  	}
...
```
In according to that example NaN() *Not a number*, but it is not logical, because the result is number - but we can`t know that the number. But it is better, then nothing. But we **the best way** is to create a agreement for that.

But problem is - How to calculate the border between angle "we can calculate the Cos" and "we cannot". Somebody call that - "bourdary conditions". For example, we can follow Java way - ignore that border:
```java
/**
* The trigonometric function <em>cos</em>. The cosine of NaN or infinity is
* NaN. This is accurate within 1 ulp, and is semi-monotonic.
*
* @param a the angle (in radians)
* @return cos(a)
*/
  public static double cos(double a)
  {
    ....
```
But In my point of view, you agree with **"We cannot calculate the Cosinus for angle 1e300"**.

**Comment by elliotchance** [comment]
I don't understand your explanation. 1e300 can be represented in a `float64` (1.8e308 would the maximum representable value), so it does not represent inf, it is just very large.

Even an extremely large number like this would still rotate around the same 2 radians and fall at a particular angle which has a defined output between 0 and 1.

**Comment by Konstantin8105** [comment]
Yes, you are right.
Result of Cos function is cannot be infinite, only from 0 to 1 and for angle 1e300 - it is also true.
And we not sure about result of Cos for angle 1e300, because we have limit  amount of decimal numbers after point.
I check that on Windows calculator, Android calcilator and sciencific portable calculator - we result of Cos(1e300) is "Error". That the reason - Why your C code return any result? Could you show the C code?

**Comment by elliotchance** [comment]
I agree that if you are calculating the cosine of such an enormous number it is more likely a bug with your application, but the mathematics still hold up.

I found this bug while creating the math tests for the https://github.com/elliotchance/c2go project: https://github.com/elliotchance/c2go/blob/master/tests/math.c#L167-L168

**Comment by ALTree** [comment]
This bug is valid, sine and cosine are defined on any real number and yes, you should be able to compute sin(1e300). But it's hard to reduce numbers this big and not terribly useful, that's why it's low priority and it hasn't been fixed (for now).

**Comment by Konstantin8105** [comment]
Just for information. I create a little java test
```java
package javaapplication2;
import static java.lang.Math.cos;
public class JavaApplication2 {
    public static void main(String[] args) {
        System.out.println(cos(1e300));
    }
}
```
The result:
```command
run:
-0.5753861119575491
BUILD SUCCESSFUL (total time: 0 seconds)
```
Java show another result.

**Comment by Konstantin8105** [comment]
We can try check Javascript.
```html
<!DOCTYPE html>
<html>
<body>
<button onclick="myFunction()">Try it</button>
<p id="demo"></p>
<script>
function myFunction() {
    document.getElementById("demo").innerHTML = Math.cos(1e300);
}
</script>
</body>
</html>
```
Please try in https://www.w3schools.com/jsref/tryit.asp?filename=tryjsref_cos , just change left code and click on button **Try it**.

**Comment by Konstantin8105** [comment]
Yes, elliotchance is right.

**Comment by elliotchance** [comment]
Actually, it seems that the cstdlib and Java/JavaScript are both giving the incorrect result:

https://www.wolframalpha.com/input/?i=cos(1e300)&rawformassumption=%22TrigRD%22+-%3E+%22R%22

cos(1e300) = -0.168214

This is more of a side note because as @ALTree said, it is difficult to reduce numbers this large.

**Comment by Konstantin8105** [comment]
Just for information.
If it is actual problem for your project/projects, then you can help for Go project https://golang.org/doc/contribute.html .

**Comment by dr2chase** [comment]
Wolfram Alpha and Java are both right.
For "cos(1.23e300)",
Wolfram Alpha computes
cos(1230000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000)
and Java computes cos(1229999999999999945619502435678791882061496502770990950045684429327960298864608335541984218516600989160291306221939122973741400364055485571676274743695192965637069768948118175959863951770799435358111025735195134313314113829815221797071926323389168215764573082356023275727273837119288529943287157489664).
or
cos(8271593907154129*2**944) (Pretend Go has Fortran's x-to-the-y operator).

Russ provided this helpful program for seeing what's going on:
https://play.golang.org/p/tI7ef6clNS
